nats/aio/transport.py

```diff
@@ -0,0 +1,231 @@
+            ), connect_timeout
+        )
+        # We keep a reference to the initial transport we used when
+        # establishing the connection in case we later upgrade to TLS
+        # after getting the first INFO message. This is in order to
+        # prevent the GC closing the socket after we send CONNECT
+        # and replace the transport.
+        #
+        # See https://github.com/nats-io/asyncio-nats/issues/43
+        self._bare_io_reader = self._io_reader = r
+        self._bare_io_writer = self._io_writer = w
+
+    async def connect_tls(
+        self,
+        uri: ParseResult,
+        ssl_context: ssl.SSLContext,
+        buffer_size: int,
+        connect_timeout: int,
+    ):
+        # loop.start_tls was introduced in python 3.7
+        # the previous method is removed in 3.9
+        if sys.version_info.minor >= 7:
+            # manually recreate the stream reader/writer with a tls upgraded transport
+            reader = asyncio.StreamReader()
+            protocol = asyncio.StreamReaderProtocol(reader)
+            transport_future = asyncio.get_running_loop().start_tls(
+                self._io_writer.transport,
+                protocol,
+                ssl_context,
+                server_hostname=uri.hostname
+            )
+            transport = await asyncio.wait_for(
+                transport_future, connect_timeout
+            )
+            writer = asyncio.StreamWriter(
+                transport, protocol, reader, asyncio.get_running_loop()
+            )
+            self._io_reader, self._io_writer = reader, writer
+        else:
+            transport = self._io_writer.transport
+            sock = transport.get_extra_info('socket')
+            if not sock:
+                # This shouldn't happen
+                raise errors.Error('nats: unable to get socket')
+
+            connection_future = asyncio.open_connection(
+                limit=buffer_size,
+                sock=sock,
+                ssl=ssl_context,
+                server_hostname=uri.hostname,
+            )
+            self._io_reader, self._io_writer = await asyncio.wait_for(
+                connection_future, connect_timeout
+            )
+
+    def write(self, payload):
+        return self._io_writer.write(payload)
+
+    def writelines(self, payload):
+        return self._io_writer.writelines(payload)
+
+    async def read(self, buffer_size: int):
+        return await self._io_reader.read(buffer_size)
+
+    async def readline(self):
+        return await self._io_reader.readline()
+
+    async def drain(self):
+        return await self._io_writer.drain()
+
+    async def wait_closed(self):
+        return await self._io_writer.wait_closed()
+
+    def close(self):
+        return self._io_writer.close()
+
+    def at_eof(self):
+        return self._io_reader.at_eof()
+
+    def __bool__(self):
+        return bool(self._io_writer) and bool(self._io_reader)
+
+
+class WebsocketTransport(Transport):
```

Comment:
how about `WebSocket` instead of `Websocket` to follow same style as aiohttp:
```suggestion
class WebSocketTransport(Transport):
```